gap/attr.gi

```diff
@@ -290,6 +290,177 @@ function(D)
   return chrom;
 end);
 
+InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
+[IsDigraph, IsDigraphColouringAlgorithm and IsDigraphColouringAlgorithmLawler],
+function(D, Lawler)
+  local n, vertices, x, s, S, i, I, s_without_I, subset_iter, induced_subgraph;
+
+  n := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif n = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+
+  vertices := DigraphVertices(D);
+  # Store current best colouring for each subset
+  x := [1 .. 2 ^ n];
+  # Empty set can be colouring with only one colour.
+  x[1] := 0;
+  subset_iter := IteratorOfCombinations(vertices);
+  # Skip the first one, which should be the empty set.
+  NextIterator(subset_iter);
+  # Iterate over all vertex subsets.
+  for s in subset_iter do
+    # Index the current subset that is being iterated over.
+    S := Sum(s, x -> 2 ^ (x - 1)) + 1;
+    x[S] := infinity;
+    # Get the subgraph induced by the vertex subset.
+    induced_subgraph := InducedSubdigraph(D, s);
+    # Iterate over the maximal independent sets of D[S]
+    for I in DigraphMaximalIndependentSets(induced_subgraph) do
+        # Create a clone of the subset that we can remove the current
+        # independent set from.
+        s_without_I := ShallowCopy(s);
+        SubtractSet(s_without_I, SetX(I,
+        # Need to relable the independent set back to the original labels.
+        x -> DigraphVertexLabel(induced_subgraph, x)));
```

Comment:
Get `DigraphVertexLabels(induced_subgraph)` and store them in a variable, then use that variable.

Reply:
Also indent this line more, since it's not a new statement.

Reply:
I don't think using `DigraphVertexLabels(induced_subgraph)` will work, as it only returns a list of the labels and not a mapping from the vertices of `induced_subgraph' to the old labels. This mapping is needed so that I can know which vertices in the original set correspond to the verticies in the maximal indepent sets of the induced subdigraph.

Reply:
I think the list of labels can be used in the same way, because the vertices are just integers.  So,
```gap
DigraphVertexLabel(induced_subgraph, x)
```
is the same as
```gap
labels := DigraphVertexLabels(induced_subgraph);
labels[x]
```
the only difference being the second one cuts down on method calls when used repeatedly.

Reply:
Ah I see, forgot that the vertices would index into that. I'll make that change now